larva/src/main/java/org/frankframework/larva/ScenarioRunner.java

```diff
@@ -15,153 +15,278 @@
 		log.info("Summary Larva run Scenario's: {} passed, {} failed. Total: {}", scenariosPassed, scenariosFailed, scenarioFiles.size());
 	}
 
-	private void runOneFile(IbisContext ibisContext, TestConfig config, File file, String currentScenariosRootDirectory, AppConstants appConstants, boolean evenStep, int waitBeforeCleanUp, String logLevel, int scenarioFileSize) {
+	private List<File> runScenariosMultithreaded(String currentScenariosRootDirectory, Map<String, List<File>> sameFolderFiles) {
+		List<File> singleThreadedScenarios = new ArrayList<>(); // Collect scenarios that should be run single threaded
+
+		// Run each scenario folder in a separate thread
+		// Not using a try-with-resources because the default awaitTermination is set on 1 day
+		ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(threads);
+		try {
+			sameFolderFiles.keySet().forEach(folder -> {
+				List<File> files = sameFolderFiles.get(folder);
+				log.debug("Starting FOLDER: {} - found: {} files", folder, files.size());
+				if (parallelBlacklistDirs.contains(folder)) {
+					log.debug("Skipping folder because found in parallel blacklist: {}", folder);
+					singleThreadedScenarios.addAll(files);
+					return;
+				}
+				executor.submit(() -> {
+					for (File file : files) {
+						runOneFile(file, currentScenariosRootDirectory);
+					}
+				});
+			});
+			// Wait for all scenarios to finish
+			executor.shutdown();
+		} finally {
+			try {
+				executor.shutdown();
+				executor.awaitTermination(15, TimeUnit.MINUTES); // Guessing a max timeout, otherwise we might hang forever
+			} catch (InterruptedException e) {
+				log.warn("Interrupted while waiting for scenario runner to finish", e);
+				Thread.currentThread().interrupt();
+			}
+		}
+		return singleThreadedScenarios;
+	}
+
+	private void runScenariosSingleThreaded(List<File> singleThreadedScenarios, String currentScenariosRootDirectory) {
+		if (singleThreadedScenarios.isEmpty()) {
+			return;
+		}
+		try {
+			config.getOut().write("<br/><h2>Starting " + singleThreadedScenarios.size() + " Single threaded Scenarios </h2>");
+		} catch (IOException ignored) {
+		}
+		multipleThreads = false; // Make output flush every scenario
+		for (File file : singleThreadedScenarios) {
+			runOneFile(file, currentScenariosRootDirectory);
+		}
+	}
+
+	// Sort property files by folder
+	private Map<String, List<File>> getSameFolderFiles(List<File> scenarioFiles, String currentScenariosRootDirectory) {
+		Map<String, List<File>> sameFolderFiles = new HashMap<>();
+		for (File file : scenarioFiles) {
+			String scenarioFolder = getScenarioFolder(file, currentScenariosRootDirectory);
+			if (sameFolderFiles.get(scenarioFolder) != null)
+				sameFolderFiles.get(scenarioFolder).add(file);
+			else {
+				List<File> files = new ArrayList<>();
+				files.add(file);
+				sameFolderFiles.put(scenarioFolder, files);
+			}
+		}
+		return sameFolderFiles;
```

Comment:
Ik kan het mishebben, maar volgens mij kan je hier in 1x het volgende doen:

```suggestion
		return scenarioFiles.stream()
				.collect(Collectors.groupingBy(scenarioFile -> getScenarioFolder(scenarioFile, currentScenariosRootDirectory)));

```